Automattic/wp-calypso
# Try/happychat as extension

This PR refactors Happychat to become a Calypso extension. By centralizing all the code in `extensions/happychat` we'll be able to leverage the idea of hybrid extensions with the goal of building a WordPress plugin out of the same code.

### To Do

- [ ] Fix linting.
- [X] Initialize Happychat on the extension.
- [ ] State:
  - [X] Happychat actions types to own extension.
  - [X] `state.happychat` to `state.extensions.happychat`.
  - [X] `state.ui.happychat` to `state.extensions.happychat.ui`.
  - [X] Middleware: convert SocketIO events + audio to data-layer handlers .
  - [X] Middleware: initialize connection. In current implementation, the SocketIO connection was started in the middleware. As a extension, HC will be loaded only in `/me/chat` section (where the extension bundle is loaded by webpack), so we don't have a method to load it in other places where Happychat is used (see UI section below). The approach in 2853458648617889033b1294075f112a02fafdf8 offers an alternative to get rid of middleware.
  - [ ] **Blocker:** HC needs the ability to [listen to all actions](https://github.com/Automattic/wp-calypso/pull/18163#discussion_r140481766). We either allow HC extension to install its own middleware or modify the data-layer to let data-handler to be specified by other things than the `action.type`.
  
  - [ ] components/happiness-support/ uses:
    - selectors: `isHappychatAvailable`.
    - UI components: `HappychatButton`, `HappychatConnection`.
  - [ ] components/marketing-survey/cancel-purchase-form/ uses:
    - UI components: `HappychatButton`.
  - [ ] layout/index.js uses:
    - selectors: `isHappychatOpen`.
    - UI components: `happychat/component`.
  - [ ] layout/footer.js uses:
    - selectors: `hasActiveHappychatSession`.
    - UI components: `HappychatButton`.
  - [ ] me/help/help-contact uses:
    - selectors: `isHappychatAvailable`.
    - actions: `openChat`, `sendChatMessage`, `sendUserInfo`.
    - UI components: `HappychatConnection`.
  - [ ] me/purchases/cancel-purchase/button.jsx
    - selectors: `isHappychatAvailable`, `hasActiveHappychatSession`.
  - [ ] me/purchases/remove-purchase/index.jsx
    - selectors: `isHappychatAvailable`, `hasActiveHappychatSession`.
    - UI components: `HappychatButton`.
  - [ ] me/sidebar: select route.
  - [ ] me/sidebar/style.scss: happychat style.
  - [ ] client/my-sites/checkout/checkout-thank-you/jetpack-thank-you-card.jsx
    - UI components: `HappychatButton`.
  - [ ] client/my-sites/checkout/checkout/payment-chat-button.jsx
    - UI components: `HappychatButton`.

### Testing instructions

* Automatic tests: `npm run test-client client/extensions/happychat`.
* As an operator, make yourself available and unavailable in the HUD (set group and locale as well). See how chat is offered or not depending on your status in several places:
  * `/help/contact`.
  * `/me/chat`.
  * `https://wordpress.com/me/purchases` - try to remove a purchase. The modal window that opens before finishing the process you'll get a "chat with us" link if there is some operator available.

Note that to start a chat session:

* an operator should be available in the HUD, with the group (WordPress.com and Jetpack should be fine) and locale of the customer account.
* the account used to open a chat in Calypso should have some paid upgrades.
